examples/rust/udp-push-pop.rs

```diff
@@ -100,168 +101,172 @@ fn close(libos: &mut LibOS, sockqd: QDesc) {
     }
 }
 
-//======================================================================================================================
-// server()
-//======================================================================================================================
+// The UDP server.
+pub struct UdpServer {
+    /// Underlying libOS.
+    libos: LibOS,
+    /// Local socket queue descriptor.
+    sockqd: QDesc,
+    /// The scatter-gather array.
+    sga: Option<demi_sgarray_t>,
+}
 
-fn server(local: SocketAddrV4) -> Result<()> {
-    let libos_name: LibOSName = match LibOSName::from_env() {
-        Ok(libos_name) => libos_name.into(),
-        Err(e) => anyhow::bail!("{:?}", e),
-    };
-    let mut libos: LibOS = match LibOS::new(libos_name) {
-        Ok(libos) => libos,
-        Err(e) => anyhow::bail!("failed to initialize libos: {:?}", e),
-    };
-    let nsends: usize = 1024;
-    let nreceives: usize = (8 * nsends) / 128;
+// Implementation of the UDP server.
+impl UdpServer {
+    pub fn new(mut libos: LibOS) -> Result<Self> {
+        // Create the local socket.
+        let sockqd: QDesc = match libos.socket(AF_INET, SOCK_DGRAM, 0) {
+            Ok(sockqd) => sockqd,
+            Err(e) => anyhow::bail!("failed to create socket: {:?}", e),
+        };
+        return Ok(Self {
+            libos,
+            sockqd,
+            sga: None,
+        });
+    }
 
-    // Setup peer.
-    let sockqd: QDesc = match libos.socket(AF_INET, SOCK_DGRAM, 0) {
-        Ok(sockqd) => sockqd,
-        Err(e) => anyhow::bail!("failed to create socket: {:?}", e),
-    };
-    match libos.bind(sockqd, local) {
-        Ok(()) => (),
-        Err(e) => {
-            // If error, close socket.
-            close(&mut libos, sockqd);
-            anyhow::bail!("bind failed: {:?}", e)
-        },
-    };
+    fn run(&mut self, local: SocketAddrV4, fill_char: u8, nsends: usize) -> Result<()> {
+        let nreceives: usize = (8 * nsends) / 128;
 
-    // Get at least nreceives.
-    for i in 0..nreceives {
-        // Receive data.
-        let qt: QToken = match libos.pop(sockqd, None) {
-            Ok(qt) => qt,
-            Err(e) => {
-                // If error, close socket.
-                close(&mut libos, sockqd);
-                anyhow::bail!("pop failed: {:?}", e)
-            },
-        };
-        let sga: demi_sgarray_t = match libos.wait(qt, None) {
-            Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_POP => unsafe { qr.qr_value.sga },
-            Ok(_) => {
-                // If error, close socket.
-                close(&mut libos, sockqd);
-                anyhow::bail!("unexpected result")
-            },
-            Err(e) => {
-                // If error, close socket.
-                close(&mut libos, sockqd);
-                anyhow::bail!("operation failed: {:?}", e)
-            },
+        match self.libos.bind(self.sockqd, local) {
```

Comment:
We should be consistent about whether we do bind in the constructor or in the run() function. I thought we had previously done it in the constructor and I think that is fine.

Reply:
We discussed offline: we rely on the fact that the drop() function is not called if the constructor (new()) fails. So, if we bind() in new(), there's a chance that it fails and the socket may not be closed because drop() was not invoked.